internal/topology/check/compatibility.go

```diff
@@ -176,10 +177,155 @@ func LocalObjectTemplateIsValid(in *clusterv1.LocalObjectTemplate, namespace str
+					field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("class"),
+					class.Class,
+					fmt.Sprintf("MachineDeployment class should be unique. MachineDeployment with class %q is defined more than once.", class.Class),
+				),
+			)
+		}
+		classes.Insert(class.Class)
+	}
+	return allErrs
+}
+
+// MachineDeploymentTopologiesAreUniqueAndDefinedInClusterClass checks that each MachineDeploymentTopology name is unique, and each class in use is defined in ClusterClass.spec.Workers.MachineDeployments.
+func MachineDeploymentTopologiesAreUniqueAndDefinedInClusterClass(desired *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if desired.Spec.Topology.Workers == nil {
+		return nil
+	}
+	if len(desired.Spec.Topology.Workers.MachineDeployments) == 0 {
+		return nil
+	}
+	// MachineDeployment clusterClass must be defined in the ClusterClass.
+	machineDeploymentClasses := classNamesFromWorkerClass(clusterClass.Spec.Workers)
+	names := sets.String{}
+	for i, md := range desired.Spec.Topology.Workers.MachineDeployments {
+		if !machineDeploymentClasses.Has(md.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "topology", "workers", "machineDeployments").Index(i),
+					md,
+					fmt.Sprintf("MachineDeployment clusterClass must be defined in the ClusterClass. MachineDeployment with clusterClass %q not found in ClusterClass %v.", md.Name, clusterClass.Name),
+				),
+			)
+		}
+		if names.Has(md.Name) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "topology", "workers", "machineDeployments").Index(i),
+					md,
+					fmt.Sprintf("MachineDeploymentTopology names should be unique. MachineDeploymentTopology with name %q is defined more than once.", md.Name),
+				),
+			)
+		}
+		names.Insert(md.Name)
+	}
+	return allErrs
+}
+
+// ClusterClassReferencesAreValid checks that each template reference in the ClusterClass is valid .
+func ClusterClassReferencesAreValid(clusterClass *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, LocalObjectTemplateIsValid(&clusterClass.Spec.Infrastructure, clusterClass.Namespace, field.NewPath("spec", "infrastructure"))...)
+	allErrs = append(allErrs, LocalObjectTemplateIsValid(&clusterClass.Spec.ControlPlane.LocalObjectTemplate, clusterClass.Namespace, field.NewPath("spec", "controlPlane"))...)
+	if clusterClass.Spec.ControlPlane.MachineInfrastructure != nil {
+		allErrs = append(allErrs, LocalObjectTemplateIsValid(clusterClass.Spec.ControlPlane.MachineInfrastructure, clusterClass.Namespace, field.NewPath("spec", "controlPlane", "machineInfrastructure"))...)
+	}
+
+	for i, mdc := range clusterClass.Spec.Workers.MachineDeployments {
+		allErrs = append(allErrs, LocalObjectTemplateIsValid(&mdc.Template.Bootstrap, clusterClass.Namespace, field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("template", "bootstrap"))...)
```

Comment:
(I wouldn't change anything in this PR, just realized it now)
Shouldn't the bootstrap template be optional? (based on that it's optional in MDs)
> 	// ConfigRef is a reference to a bootstrap provider-specific resource
>	// that holds configuration details. The reference is optional to
>	// allow users/operators to specify Bootstrap.DataSecretName without
>	// the need of a controller.

Or is that something not supported by ClusterClass?

Reply:
cc @fabriziopandini Is this something we have to think about / create an issue for?

Reply:
I don't think so.
it is kind of impossible automatically creating clusters without a bootstrap provider, given that it is impractical or impossible to provide bootstrap data secrets at ClusterClass level

Reply:
Okay sounds good. Pretty relieved that we don't have to support optional BootstrapTemplates